PLEASE FIX OR REPORT | Unexpected exception thrown for events-async-8

Incident [reported on discord](https://discord.com/channels/697535391594446898/697539289042649190/1058599473959403611).

Logs are [here](https://logpaste.com/HEnuwD9P). The relevant part copied below. Looks like not a lot to go on. May be related to a previous error reported a minute earlier `ERROR - Failed to process event queue for client 921726489`.

Use reports that the incident was isolated and everything is running fine since.

```
consensus_1                  | 2022-12-31 04:00:49.151 ERROR - PLEASE FIX OR REPORT | Unexpected exception thrown for events-async-8
consensus_1                  | java.util.concurrent.CompletionException: java.lang.IllegalStateException: s=OPEN,api=BLOCKED,sc=false,e=null
consensus_1                  | at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:146) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.whenComplete(Unknown Source) ~[?:?]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:603) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:33) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:143) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$33(SafeFuture.java:422) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:146) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.CompletableFuture.whenComplete(Unknown Source) ~[?:?]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:603) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:33) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:143) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateTo(SafeFuture.java:316) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:119) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
consensus_1                  | at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
consensus_1                  | at java.lang.Thread.run(Unknown Source) ~[?:?]
consensus_1                  | Caused by: java.lang.IllegalStateException: s=OPEN,api=BLOCKED,sc=false,e=null
consensus_1                  | at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:813) ~[jetty-server-11.0.12.jar:11.0.12]
consensus_1                  | at org.eclipse.jetty.server.HttpOutput.print(HttpOutput.java:1126) ~[jetty-server-11.0.12.jar:11.0.12]
consensus_1                  | at org.eclipse.jetty.server.HttpOutput.print(HttpOutput.java:1053) ~[jetty-server-11.0.12.jar:11.0.12]
consensus_1                  | at io.javalin.http.sse.Emitter.write(Emitter.kt:42) ~[javalin-5.2.0.jar:5.2.0]
consensus_1                  | at io.javalin.http.sse.Emitter.emit(Emitter.kt:34) ~[javalin-5.2.0.jar:5.2.0]
consensus_1                  | at io.javalin.http.sse.SseClient.sendComment(SseClient.kt:81) ~[javalin-5.2.0.jar:5.2.0]
consensus_1                  | at tech.pegasys.teku.beaconrestapi.handlers.v1.events.EventSubscriber.lambda$keepAlive$4(EventSubscriber.java:147) ~[teku-data-beaconrestapi-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.fromRunnable(SafeFuture.java:155) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAfterDelay$1(AsyncRunner.java:32) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:74) ~[teku-infrastructure-async-22.12.0.jar:22.12.0]
consensus_1                  | ... 4 more
```

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
@@ -156,8 +156,7 @@ public class EventSubscriber {
                 final Throwable rootException = Throwables.getRootCause(throwable);
                 if (rootException instanceof IllegalStateException) {
                   // this might fail due to https://github.com/eclipse/jetty.project/issues/4461 but
-                  // ultimately
-                  // isn't important because it has a retry mechanism
+                  // ultimately isn't important because it has a retry mechanism
                   LOG.warn(
                       "Error sending keep-alive SSE comment. Retrying in {} seconds.",
                       retryInSeconds);

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
@@ -157,7 +157,7 @@ public class EventSubscriber {
                 if (rootException instanceof IllegalStateException) {
                   // this might fail due to https://github.com/eclipse/jetty.project/issues/4461 but
                   // ultimately isn't important because it has a retry mechanism
-                  LOG.warn(
+                  LOG.debug(
                       "Error sending keep-alive SSE comment. Retrying in {} seconds.",
                       retryInSeconds);
                   return null;

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
@@ -14,6 +14,7 @@
 package tech.pegasys.teku.beaconrestapi.handlers.v1.events;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.base.Throwables;
 import io.javalin.http.sse.SseClient;
 import java.io.ByteArrayInputStream;
 import java.time.Duration;
@@ -138,6 +139,7 @@ public class EventSubscriber {
   }
 
   private void keepAlive() {
+    final int retryInSeconds = 30;
     if (!stopped.get()) {
       asyncRunner
           .runAfterDelay(
@@ -147,9 +149,28 @@ public class EventSubscriber {
                   sseClient.sendComment("");
                 }
               },
-              Duration.ofSeconds(30))
+              Duration.ofSeconds(retryInSeconds))
           .alwaysRun(this::keepAlive)
-          .ifExceptionGetsHereRaiseABug();
+          .exceptionally(
+              throwable -> {
+                final Throwable rootException = Throwables.getRootCause(throwable);
+                if (rootException instanceof IllegalStateException) {
+                  // this might fail due to https://github.com/eclipse/jetty.project/issues/4461 but
+                  // ultimately
+                  // isn't important because it has a retry mechanism
+                  LOG.warn(
+                      "Error sending keep-alive SSE comment. Retrying in {} seconds.",
+                      retryInSeconds);
+                  return null;
+                }
+
+                if (throwable instanceof RuntimeException) {
+                  throw (RuntimeException) throwable;
+                } else {
+                  throw new RuntimeException(
+                      "Unexpected error when sending keep-alive SSE comment.", throwable);
+                }
+              });
     }
   }
 

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
@@ -138,6 +138,7 @@ public class EventSubscriber {
                     "Failed to process event queue for client " + sseClient.hashCode(), error));
   }
 
+  @SuppressWarnings("FutureReturnValueIgnored")
   private void keepAlive() {
     final int retryInSeconds = 30;
     if (!stopped.get()) {

